README
= Name

Jabber::Simple - An extremely easy-to-use Jabber client library.

= Synopsis

  # Send a message to a friend, asking for authorization if necessary:
  im = Jabber::Simple.new("user@example.com", "password")
  im.deliver("friend@example.com", "Hey there friend!")

  # Get received messages and print them out to the console:
  im.received_messages { |msg| puts msg.body if msg.type == :chat }

  # Send an authorization request to a user:
  im.add("friend@example.com")

  # Get presence updates from your friends, and print them out to the console:
  # (admittedly, this one needs some work)
  im.presence_updates do |update|
    from     = update[0].jid.strip.to_s
    status   = update[2].status
    presence = update[2].show
    puts "#{from} went #{presence}: #{status}"
  end

  # Remove a user from your contact list:
  im.remove("unfriendly@example.com")

  # See the Jabber::Simple documentation for more information.

= Description

Jabber::Simple is intended to make Jabber client programming dead simple. XMPP,
the Jabber protocol, is extremely powerful but also carries a steep learning
curve. This library exposes only the most common tasks, and does so in a way
that is familiar to users of traditional instant messenger clients.

= Known Issues

* None. If you'd like additional functionality, please contact the developer!

= Copyright

Jabber::Simple - An extremely easy-to-use Jabber client library.
Copyright 2006-2008 Blaine Cook <romeda@gmail.com>.

Jabber::Simple is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

Jabber::Simple is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Jabber::Simple; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA


CHANGELOG
xmpp4r-simple (0.8.9)

  [ Alexander Pervakov ]
  * Changed reconnect timeout according to RFC 6120.

  [ Emmett Shear, thelinuxlich, Alexander Pervakov ]
  * Adding connection options (host, port, allow_tls).

  [ David Rice ]
  * Ruby 1.9 compatibility fix.

-- Alexander Pervakov <frost.nzcr4@jagmort.com> Monday, 04 May 2015 15:09:00 +0300

xmpp4r-simple (0.8.8)

  [ Blaine Cook ]
  * Add xmpp4r-simple.gemspec, remove gem-related code from Rakefile

-- Blaine Cook <romeda@gmail.com> Wed, 30 Jul 2008 16:03:00 -0800

xmpp4r-simple (0.8.7)

  [ Blaine Cook ]
  * If using Jabber::Simple in a DRb environment, the Jabber::Simple object
   	will execute in the DRb server's environment to improve efficiency.
  * Jabber::Simple will now recover gracefully from server-side disconnects.
  * Updated Rakefile to pass tests on firebrigade (http://firebrigade.seattlerb.org)

-- Blaine Cook <romeda@gmail.com> Wed, 31 Jan 2007 17:15:56 -0800

xmpp4r-simple (0.8.6)

  [ Blaine Cook ]
  * Catch a rare exception that's thrown when we are disconnecting.

-- Blaine Cook <romeda@gmail.com> Fri, 26 Jan 2007 11:45:23 -0800


xmpp4r-simple (0.8.5)

  [ Blaine Cook ]
  * Accept Jabber::Message messages as arguments to Jabber::Simple#deliver

-- Blaine Cook <romeda@gmail.com>  Mon, 22 Jan 2007 07:23:09 -0800


xmpp4r-simple (0.8.4)

  [ Blaine Cook ]
  * Catch broken connections and attempt to reconnect 3 times.

-- Blaine Cook <romeda@gmail.com>  Fri, 23 Dec 2006 00:12:09 -0800


xmpp4r-simple (0.8.3)

  [ Blaine Cook ]
  * Update presence_updates to only store one presence_update per user.
		Changes methods, will break code that uses presence_updates if not updated
		correspondingly, check the documenation for the new semantics.

-- Blaine Cook <romeda@gmail.com>  Thu, 07 Dec 2006 12:45:52 -0800


xmpp4r-simple (0.8.2)

  [ Blaine Cook ]
  * Add presence_updates?, received_messages?, and new_subscriptions? methods.

-- Blaine Cook <romeda@gmail.com>  Wed, 06 Dec 2006 09:40:28 -0800


xmpp4r-simple (0.8.1)

  [ Blaine Cook ]
  * Make the deferred_delivery method public

-- Blaine Cook <romeda@gmail.com>  Tue, 05 Dec 2006 17:39:14 -0800


xmpp4r-simple (0.8.0)

  [ Blaine Cook ]
  * initial import

-- Blaine Cook <romeda@gmail.com>  Wed,  08 Nov 2006 20:42:42 -0800


test/test_xmpp4r_simple.rb
# Jabber::Simple - An extremely easy-to-use Jabber client library.
# Copyright 2006 Blaine Cook <blaine@obvious.com>, Obvious Corp.
# 
# Jabber::Simple is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# Jabber::Simple is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with Jabber::Simple; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
$:.unshift "#{File.dirname(__FILE__)}/../lib"

require 'test/unit'
require 'timeout'
require 'xmpp4r-simple'

class JabberSimpleTest < Test::Unit::TestCase

  def setup
    @@connections ||= {}

    if @@connections.include?(:client1)
      @client1 = @@connections[:client1]
      @client2 = @@connections[:client2]
      @client1.accept_subscriptions = true
      @client2.accept_subscriptions = true
      @jid1_raw = @@connections[:jid1_raw]
      @jid2_raw = @@connections[:jid2_raw]
      @jid1 = @jid1_raw.strip.to_s
      @jid2 = @jid2_raw.strip.to_s
      return true
    end

    logins = []
    begin
      logins = File.readlines(File.expand_path("~/.xmpp4r-simple-test-config")).map! { |login| login.split(" ") }
      raise StandardError unless logins.size == 2
    rescue => e
      puts "\nConfiguration Error!\n\nYou must make available two unique Jabber accounts in order for the tests to pass."
      puts "Place them in ~/.xmpp4r-simple-test-config, one per line like so:\n\n"
      puts "user1@example.com/res password"
      puts "user2@example.com/res password\n\n"
      raise e
    end

    @@connections[:client1] = Jabber::Simple.new(*logins[0])
    @@connections[:client2] = Jabber::Simple.new(*logins[1])

    @@connections[:jid1_raw] = Jabber::JID.new(logins[0][0])
    @@connections[:jid2_raw] = Jabber::JID.new(logins[1][0])

    # Force load the client and roster, just to be safe.
    @@connections[:client1].roster
    @@connections[:client2].roster

    # Re-run this method to setup the local instance variables the first time.
    setup
  end

  def test_ensure_the_jabber_clients_are_connected_after_setup
    assert @client1.client.is_connected?
    assert @client2.client.is_connected?
  end

  def test_inspect_should_be_custom
    assert_equal "Jabber::Simple #{@jid1_raw}", @client1.inspect
  end

  def test_inspect_contact_should_be_custom
    assert_equal "Jabber::Contact #{@jid2}", @client1.contacts(@jid2).inspect
  end

  def test_remove_users_from_our_roster_should_succeed
    @client2.remove(@jid1)
    @client1.remove(@jid2)

    sleep 3

    assert_equal false, @client1.subscribed_to?(@jid2)
    assert_equal false, @client2.subscribed_to?(@jid1)
  end

  def test_add_users_to_our_roster_should_succeed_with_automatic_approval
    @client1.remove(@jid2)
    @client2.remove(@jid1)

    assert_before 60 do
      assert_equal false, @client1.subscribed_to?(@jid2)
      assert_equal false, @client2.subscribed_to?(@jid1)
    end

    @client1.new_subscriptions
    @client1.add(@jid2)

    assert_before 60 do
      assert @client1.subscribed_to?(@jid2)
      assert @client2.subscribed_to?(@jid1)
    end

    new_subscriptions = @client1.new_subscriptions
    assert_equal 1, new_subscriptions.size
    assert_equal @jid2, new_subscriptions[0][0].jid.strip.to_s
  end

  def test_sent_message_should_be_received
    # First clear the client's message queue, just in case.
    assert_kind_of Array, @client2.received_messages

    # Next ensure that we're not subscribed, so that we can test the deferred message queue.
    @client1.remove(@jid2)
    @client2.remove(@jid1)
    sleep 2

    # Deliver the messages; this should be received by the other client.
    @client1.deliver(@jid2, "test message")

    sleep 2

    # Fetch the message; allow up to ten seconds for the delivery to occur.
    messages = []
    begin
      Timeout::timeout(20) {
        loop do
          messages = @client2.received_messages
          break unless messages.empty?
          sleep 1
        end
      }
    rescue Timeout::Error
      flunk "Timeout waiting for message"
    end

    # Ensure that the message was received intact.
    assert_equal @jid1, messages.first.from.strip.to_s
    assert_equal "test message", messages.first.body
  end

  def test_presence_updates_should_be_received

    @client2.add(@client1)

    assert_before(60) { assert @client2.subscribed_to?(@jid1) }
    assert_before(60) { assert_equal 0, @client2.presence_updates.size }

    @client1.status(:away, "Doing something else.")
    
    new_statuses = []
    assert_before(60) do
      new_statuses = @client2.presence_updates
      assert_equal 1, new_statuses.size
    end

    new_status = new_statuses.first
    assert_equal @jid1, new_status[0].jid.strip.to_s
    assert_equal "Doing something else.", new_status[2].status
    assert_equal :away, new_status[2].show
  end

  def test_disable_auto_accept_subscription_requests
    @client1.remove(@jid2)
    @client2.remove(@jid1)

    assert_before(60) do
      assert_equal false, @client1.subscribed_to?(@jid2)
      assert_equal false, @client2.subscribed_to?(@jid1)
    end

    @client1.accept_subscriptions = false
    assert_equal false, @client1.accept_subscriptions?
    
    assert_before(60) { assert_equal 0, @client1.subscription_requests.size }

    @client2.add(@jid1)

    new_subscription_requests = []
    assert_before(60) do
      new_subscription_requests = @client1.subscription_requests
      assert_equal 1, new_subscription_requests.size
    end

    new_subscription = new_subscription_requests.first
    assert_equal @jid2, new_subscription[0].jid.strip.to_s
    assert_equal :subscribe, new_subscription[1].type
  end

  def test_automatically_reconnect
    @client1.client.close

    assert_before 60 do
      assert_equal false, @client1.connected?
    end

    # empty client 2's received message queue.
    @client2.received_messages

    @client1.deliver(@jid2, "Testing")

    assert_before(60) { assert @client1.connected? }
    assert @client1.roster.instance_variable_get('@stream').is_connected?
    assert_before(60) { assert_equal 1, @client2.received_messages.size }
  end

  def test_disconnect_doesnt_allow_auto_reconnects
    @client1.disconnect

    assert_equal false, @client1.connected?
    
    assert_raises Jabber::ConnectionError do
      @client1.deliver(@jid2, "testing")
    end

    @client1.reconnect
  end

  private

  def assert_before(seconds, &block)
    error = nil
    begin
      Timeout::timeout(seconds) {
        begin
          yield
        rescue => e
          error = e
          sleep 0.5
          retry
        end
      }
    rescue Timeout::Error
      raise error
    end
  end

end


test/xmpp4r-simple-test.rb
require 'test/unit'
require 'timeout'
require '../lib/xmpp4r-simple'

class JabberSimpleTest < Test::Unit::TestCase

  def setup
    return true if @initialized
    @initialized = true

    logins = []
    begin
      logins = File.readlines(File.expand_path("~/.xmpp4r-simple-test-config")).map! { |login| login.split(" ") }
      raise StandardError unless logins.size == 2
    rescue => e
      puts "\nConfiguration Error!\n\nYou must make available two unique Jabber accounts in order for the tests to pass."
      puts "Place them in ~/.xmpp4r-simple-test-config, one per line like so:\n\n"
      puts "user1@example.com/res password"
      puts "user2@example.com/res password\n\n"
      raise e
    end

    @client1 = Jabber::Simple.new(*logins[0])
    @client2 = Jabber::Simple.new(*logins[1])

    @jid1 = Jabber::JID.new(logins[0][0]).strip.to_s
    @jid2 = Jabber::JID.new(logins[1][0]).strip.to_s

    # Force load the client and roster, just to be safe.
    @client1.roster
    @client2.roster
  end

  def test_ensure_the_jabber_clients_are_connected_after_setup
    assert @client1.client.is_connected?
    assert @client2.client.is_connected?
  end

  def test_remove_users_from_our_roster_should_succeed
    @client2.remove(@jid1)
    @client1.remove(@jid2)

    sleep 3

    assert_equal false, @client1.subscribed_to?(@jid2)
    assert_equal false, @client2.subscribed_to?(@jid1)
  end

  def test_add_users_to_our_roster_should_succeed_with_automatic_approval
    @client1.remove(@jid2)
    @client2.remove(@jid1)

    assert_before 60 do
      assert_equal false, @client1.subscribed_to?(@jid2)
      assert_equal false, @client2.subscribed_to?(@jid1)
    end

    sleep 2
    @client1.add(@jid2)

    assert_before 60 do
      assert @client1.subscribed_to?(@jid2)
      assert @client2.subscribed_to?(@jid1)
    end
  end

  def test_sent_message_should_be_received
    # First clear the client's message queue, just in case.
    assert_kind_of Array, @client2.received_messages

    # Next ensure that we're not subscribed, so that we can test the deferred message queue.
    @client1.remove(@jid2)
    @client2.remove(@jid1)
    sleep 2

    # Deliver the messages; this should be received by the other client.
    @client1.deliver(@jid2, "test message")

    sleep 2

    # Fetch the message; allow up to ten seconds for the delivery to occur.
    messages = []
    begin
      Timeout::timeout(10) {
        loop do
          messages = @client2.received_messages
          break unless messages.empty?
          sleep 1
        end
      }
    rescue Timeout::Error
      flunk "Timeout waiting for message"
    end

    # Ensure that the message was received intact.
    assert_equal @jid1, messages.first.from.strip.to_s
    assert_equal "test message", messages.first.body
  end

  private

  def assert_before(seconds, &block)
    error = nil
    begin
      Timeout::timeout(seconds) {
        begin
          yield
        rescue => e
          error = e
          sleep 0.5
          retry
        end
      }
    rescue Timeout::Error
      raise error
    end
  end

end


lib/xmpp4r-simple.rb
# Jabber::Simple - An extremely easy-to-use Jabber client library.
# Copyright 2006 Blaine Cook <blaine@obvious.com>, Obvious Corp.
# 
# Jabber::Simple is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# Jabber::Simple is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with Jabber::Simple; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

require 'rubygems'
require 'xmpp4r'
require 'xmpp4r/roster'
require 'xmpp4r/vcard'

module Jabber

  class ConnectionError < StandardError #:nodoc:
  end
  
  class Contact #:nodoc:

    include DRb::DRbUndumped if defined?(DRb::DRbUndumped)

    def initialize(client, jid)
      @jid = jid.respond_to?(:resource) ? jid : JID.new(jid)
      @client = client
    end

    def inspect
      "Jabber::Contact #{jid.to_s}"
    end

    def subscribed?
      [:to, :both].include?(subscription)
    end

    def subscription
      roster_item && roster_item.subscription
    end

    def ask_for_authorization!
      subscription_request = Presence.new.set_type(:subscribe)
      subscription_request.to = jid
      client.send!(subscription_request)
    end
    
    def unsubscribe!
      unsubscription_request = Presence.new.set_type(:unsubscribe)
      unsubscription_request.to = jid
      client.send!(unsubscription_request)
      client.send!(unsubscription_request.set_type(:unsubscribed))
    end

    def jid(bare=true)
      bare ? @jid.strip : @jid
    end

    private

    def roster_item
      client.roster.items[jid]
    end
   
    def client
      @client
    end
  end

  class Simple

    include DRb::DRbUndumped if defined?(DRb::DRbUndumped)

    # Create a new Jabber::Simple client. You will be automatically connected
    # to the Jabber server and your status message will be set to the string
    # passed in as the status_message argument.
    #
    # If you'd like to connect to a different talk server than the one which would
    # be guessed from your jid, you may provide a server. For example, to connect
    # to the gmail talk servers with a jid that doesn't end in @gmail.com, just provide
    # options {:host => 'talk.l.google.com'}.
    #
    # jabber = Jabber::Simple.new("me@example.com", "password", "Chat with me - Please!", "Available", {:host => "host-not-from-jid.tld", :port => 64321})
    def initialize(jid, password, status = nil, status_message = "Available", opts= {})
      @jid = jid
      @password = password
      @disconnected = false
      @opts = {
        :host => opts[:host],  # nil by default as in xmpp4r.
        :port => !opts[:port].nil? ? opts[:port] : 5222,  # xmpp4r default.
        :allow_tls => !opts[:allow_tls].nil? ? opts[:allow_tls] : true  # xmpp4r default.
      }
      status(status, status_message)
      start_deferred_delivery_thread
    end

    def inspect #:nodoc:
      "Jabber::Simple #{@jid}"
    end
    
    # Send a message to jabber user jid.
    #
    # Valid message types are:
    # 
    #   * :normal (default): a normal message.
    #   * :chat: a one-to-one chat message.
    #   * :groupchat: a group-chat message.
    #   * :headline: a "headline" message.
    #   * :error: an error message.
    #
    # If the recipient is not in your contacts list, the message will be queued
    # for later delivery, and the Contact will be automatically asked for
    # authorization (see Jabber::Simple#add).
    #
    # message should be a string or a valid Jabber::Message object. In either case,
    # the message recipient will be set to jid.
    def deliver(jid, message, type=:chat)
      contacts(jid) do |friend|
        unless subscribed_to? friend
          add(friend.jid)
          return deliver_deferred(friend.jid, message, type)
        end
        if message.kind_of?(Jabber::Message)
          msg = message
          msg.to = friend.jid
        else
          msg = Message.new(friend.jid)
          msg.type = type
          msg.body = message
        end
        send!(msg)
      end
    end

    # Set your presence, with a message.
    #
    # Available values for presence are:
    # 
    #   * nil: online.
    #   * :chat: free for chat.
    #   * :away: away from the computer.
    #   * :dnd: do not disturb.
    #   * :xa: extended away.
    #
    # It's not possible to set an offline status - to do that, disconnect! :-)
    def status(presence, message)
      @presence = presence
      @status_message = message
      stat_msg = Presence.new(@presence, @status_message)
      send!(stat_msg)
    end

    # Ask the users specified by jids for authorization (i.e., ask them to add
    # you to their contact list). If you are already in the user's contact list,
    # add() will not attempt to re-request authorization. In order to force
    # re-authorization, first remove() the user, then re-add them.
    #
    # Example usage:
    # 
    #   jabber_simple.add("friend@friendosaurus.com")
    #
    # Because the authorization process might take a few seconds, or might
    # never happen depending on when (and if) the user accepts your
    # request, results are placed in the Jabber::Simple#new_subscriptions queue.
    def add(*jids)
      contacts(*jids) do |friend|
        next if subscribed_to? friend
        friend.ask_for_authorization!
      end
    end

    # Remove the jabber users specified by jids from the contact list.
    def remove(*jids)
      contacts(*jids) do |unfriend|
        unfriend.unsubscribe!
      end
    end

    # Returns true if this Jabber account is subscribed to status updates for
    # the jabber user jid, false otherwise.
    def subscribed_to?(jid)
      contacts(jid) do |contact|
        return contact.subscribed?
      end
    end

    # If contacts is a single contact, returns a Jabber::Contact object
    # representing that user; if contacts is an array, returns an array of
    # Jabber::Contact objects.
    #
    # When called with a block, contacts will yield each Jabber::Contact object
    # in turn. This is mainly used internally, but exposed as an utility
    # function.
    def contacts(*contacts, &block)
      @contacts ||= {}
      contakts = []
      contacts.each do |contact|
        jid = contact.to_s
        unless @contacts[jid]
          @contacts[jid] = contact.respond_to?(:ask_for_authorization!) ? contact : Contact.new(self, contact)
        end
        yield @contacts[jid] if block_given?
        contakts << @contacts[jid]
      end
      contakts.size > 1 ? contakts : contakts.first
    end

    # Returns true if the Jabber client is connected to the Jabber server,
    # false otherwise.
    def connected?
      @client ||= nil
      connected = @client.respond_to?(:is_connected?) && @client.is_connected?
      return connected
    end

    # Returns an array of messages received since the last time
    # received_messages was called. Passing a block will yield each message in
    # turn, allowing you to break part-way through processing (especially
    # useful when your message handling code is not thread-safe (e.g.,
    # ActiveRecord).
    #
    # e.g.:
    #
    #   jabber.received_messages do |message|
    #     puts "Received message from #{message.from}: #{message.body}"
    #   end
    def received_messages(&block)
      dequeue(:received_messages, &block)
    end

    # Returns true if there are unprocessed received messages waiting in the
    # queue, false otherwise.
    def received_messages?
      !queue(:received_messages).empty?
    end

    # Returns an array of presence updates received since the last time
    # presence_updates was called. Passing a block will yield each update in
    # turn, allowing you to break part-way through processing (especially
    # useful when your presence handling code is not thread-safe (e.g.,
    # ActiveRecord).
    #
    # e.g.:
    #
    #   jabber.presence_updates do |friend, new_presence|
    #     puts "Received presence update from #{friend}: #{new_presence}"
    #   end
    def presence_updates(&block)
      updates = []
      @presence_mutex.synchronize do
        dequeue(:presence_updates) do |friend|
          presence = @presence_updates[friend]
          next unless presence
          new_update = [friend, presence[0], presence[1]]
          yield new_update if block_given?
          updates << new_update
          @presence_updates.delete(friend)
        end
      end
      return updates
    end
    
    # Returns true if there are unprocessed presence updates waiting in the
    # queue, false otherwise.
    def presence_updates?
      !queue(:presence_updates).empty?
    end

    # Returns an array of subscription notifications received since the last
    # time new_subscriptions was called. Passing a block will yield each update
    # in turn, allowing you to break part-way through processing (especially
    # useful when your subscription handling code is not thread-safe (e.g.,
    # ActiveRecord).
    #
    # e.g.:
    #
    #   jabber.new_subscriptions do |friend, presence|
    #     puts "Received presence update from #{friend.to_s}: #{presence}"
    #   end
    def new_subscriptions(&block)
      dequeue(:new_subscriptions, &block)
    end
    
    # Returns true if there are unprocessed presence updates waiting in the
    # queue, false otherwise.
    def new_subscriptions?
      !queue(:new_subscriptions).empty?
    end
    
    # Returns an array of subscription notifications received since the last
    # time subscription_requests was called. Passing a block will yield each update
    # in turn, allowing you to break part-way through processing (especially
    # useful when your subscription handling code is not thread-safe (e.g.,
    # ActiveRecord).
    #
    # e.g.:
    #
    #   jabber.subscription_requests do |friend, presence|
    #     puts "Received presence update from #{friend.to_s}: #{presence}"
    #   end
    def subscription_requests(&block)
      dequeue(:subscription_requests, &block)
    end
 
    # Returns true if auto-accept subscriptions (friend requests) is enabled
    # (default), false otherwise.
    def accept_subscriptions?
      @accept_subscriptions = true if @accept_subscriptions.nil?
      @accept_subscriptions
    end

    # Change whether or not subscriptions (friend requests) are automatically accepted.
    def accept_subscriptions=(accept_status)
      @accept_subscriptions = accept_status
    end
    
    # Direct access to the underlying Roster helper.
    def roster
      return @roster if @roster
      self.roster = Roster::Helper.new(client)
    end

    # Direct access to the underlying Jabber client.
    def client
      connect!() unless connected?
      @client
    end
    
    # Send a Jabber stanza over-the-wire.
    def send!(msg)
      attempts = 0
      begin
        attempts += 1
        client.send(msg)
      rescue Errno::EPIPE, IOError => e
        sleep 1
        disconnect
        reconnect
        retry unless attempts > 3
        raise e
      rescue Errno::ECONNRESET => e
        # @see http://xmpp.org/rfcs/rfc6120.html#tcp-reconnect XMPP: Core - Reconnection
        sleep 1 + rand(60) + (attempts - 1) * 20
        disconnect
        reconnect
        retry unless attempts > 3
        raise e
      end
    end

    # Use this to force the client to reconnect after a force_disconnect.
    def reconnect
      @disconnected = false
      connect!
    end

    # Use this to force the client to disconnect and not automatically
    # reconnect.
    def disconnect
      disconnect!
    end
    
    # Queue messages for delivery once a user has accepted our authorization
    # request. Works in conjunction with the deferred delivery thread.
    #
    # You can use this method if you want to manually add friends and still
    # have the message queued for later delivery.
    def deliver_deferred(jid, message, type)
      msg = {:to => jid, :message => message, :type => type}
      queue(:pending_messages) << [msg]
    end

    private

    def client=(client)
      self.roster = nil # ensure we clear the roster, since that's now associated with a different client.
      @client = client
    end

    def roster=(new_roster)
      @roster = new_roster
    end

    def connect!
      raise ConnectionError, "Connections are disabled - use Jabber::Simple::force_connect() to reconnect." if @disconnected
      # Pre-connect
      @connect_mutex ||= Mutex.new

      # don't try to connect if another thread is already connecting.
      return if @connect_mutex.locked?

      @connect_mutex.lock
      disconnect!(false) if connected?

      # Connect
      jid = JID.new(@jid)
      my_client = Client.new(@jid)
      my_client.allow_tls = @opts[:allow_tls]
      my_client.connect @opts[:host], @opts[:port]
      my_client.auth(@password)
      self.client = my_client

      # Post-connect
      register_default_callbacks
      status(@presence, @status_message)
      @connect_mutex.unlock
    end

    def disconnect!(auto_reconnect = true)
      if client.respond_to?(:is_connected?) && client.is_connected?
        begin
          client.close
        rescue Errno::EPIPE, IOError => e
          # probably should log this.
          nil
        end
      end
      client = nil
      @disconnected = auto_reconnect
    end

    def register_default_callbacks
      client.add_message_callback do |message|
        queue(:received_messages) << message unless message.body.nil?
      end

      roster.add_subscription_callback do |roster_item, presence|
        if presence.type == :subscribed
          queue(:new_subscriptions) << [roster_item, presence]
        end
      end

      roster.add_subscription_request_callback do |roster_item, presence|
        if accept_subscriptions?
          roster.accept_subscription(presence.from) 
        else
          queue(:subscription_requests) << [roster_item, presence]
        end
      end

      @presence_updates = {}
      @presence_mutex = Mutex.new
      roster.add_presence_callback do |roster_item, old_presence, new_presence|
        simple_jid = roster_item.jid.strip.to_s
        presence = case new_presence.type
                   when nil then new_presence.show || :online
                   when :unavailable then :unavailable
                   else
                     nil
                   end

        if presence && @presence_updates[simple_jid] != presence
          queue(:presence_updates) << simple_jid
          @presence_mutex.synchronize { @presence_updates[simple_jid] = [presence, new_presence.status] }
        end
      end
    end

    # This thread facilitates the delivery of messages to users who haven't yet
    # accepted an invitation from us. When we attempt to deliver a message, if
    # the user hasn't subscribed, we place the message in a queue for later
    # delivery. Once a user has accepted our authorization request, we deliver
    # any messages that have been queued up in the meantime.
    def start_deferred_delivery_thread #:nodoc:
      Thread.new {
        loop {
          messages = [queue(:pending_messages).pop].flatten
          messages.each do |message|
            if subscribed_to?(message[:to])
              deliver(message[:to], message[:message], message[:type])
            else
              queue(:pending_messages) << message
            end
          end
        }
      }
    end

    def queue(queue)
      @queues ||= Hash.new { |h,k| h[k] = Queue.new }
      @queues[queue]
    end

    def dequeue(queue, non_blocking = true, max_items = 100, &block)
      queue_items = []
      max_items.times do
        queue_item = queue(queue).pop(non_blocking) rescue nil
        break if queue_item.nil?
        queue_items << queue_item
        yield queue_item if block_given?
      end
      queue_items
    end
  end
end

true
